### PR TITLE
add generics

### PR DESCRIPTION
**Pull request by arnaucube**
Doing this through a branch & PR so we can review it better and evaluate before deciding if to use it or not.

Some notes:
I thought that is comfortable to enjoy `F` in the CCS impl instead of `C::ScalarField` which is longer. The problem is that later for LCCCS & CCCS, due the Commitment, they need to work over the curve group `C: CurveGroup`, thus all structs & methods that until now were working on `Fr` and that now would be `F`, it will need to be `C::ScalarField`.
But in the current version, `CCS` is 'protected' from this and can keep enjoying working with `F` (instead of `C::ScalarField`).
This comes at the 'cost' of in the methods `to_lcccs` and `to_cccs` we need to 'convert' the `CCS<F>` into `CCS<C::ScalarField>`. Is not a big deal, but it would be more clean if CCS was already on `C` instead of `F`, even that CCS does not need any curve group stuff except for `C::ScalarField`.

Let me know what do you think, I'm currently slightly inclined in favour of migrating CCS to `C` also, as the code involved is few stuff compared to the current version of 'hacking' the `CCS<F>` to `CCS<C::ScalarField>`, but for the moment I left it in that way so we can evaluate it better.

_

For the rest, I think that the verbosity added because of generics is still manageable for the benefit of having the multifolding implementation with generics (as the rest of arkworks & hyperplonk libs).

### CODE REVIEW

**Comment by asn-d6** [comment]
I don't have a strong opinion here.

On the plus side, I agree that it's more useful as generic. It's also the standard wrt arkworks libraries.
On the negative side, since this is a PoC codebase and we will likely port to Nova in the future, generics add a small readability and porting complexity that is not necessary for a PoC.

tl;dr I think if we keep this as a PoC, we can maximize for readability and learning and not use generics. If we think this might be used in other contexts in the future, we should use generics.

All in all, no strong opinion and I'm fine either way.

**Comment by arnaucube** [comment]
When I created the PR I had no strong opinions either, and was more a 'exploratory' path.
But after meditating about it, I'm now more inclined towards adding generics because:
at some point, the multifolding will need to be done through a cycle of curves (in the same way than in Nova), thus that will require generics. We can delay the generics thing, but at some point before porting this repo to Nova, or when we port it, the generics will need to be added. Thus, we can already do it now.
The only objection that I can think against moving to generics is the extra complexity in the code, but as can be seen in this PR, the difference is quite small.

In the first message of this PR I commented on the attempt to keep the `CCS` struct & methods working with `F` instead of with `C::ScalarField` just because of less characters. But now after reviewing the Nova code for the R1CS instances, in which it uses directly the curve group (G), makes more sense to get rid of the extra methods to convert from `CCS<F>` to `CCS<C>`, and directly use `CCS<C>` everywhere. Will do a force push to update the branch of this PR with this change.

**Comment by asn-d6** [comment]
Sounds good. Merging.
